Drivers/Src/Display.c

```diff
@@ -1,325 +1,141 @@
-    char buf[8];
-    BSP_UART_Read(DISP_OUT, buf);
-    return (IsNextionFailure(buf[0])) ? ERROR : SUCCESS;
-}
+#define DISP_OUT UART_3
+static const char *TERMINATOR = "\xff\xff\xff";
 
-/*
- * Sets a component's visiblity
- */
-ErrorStatus setComponentVisibility(enum CommandString_t comp, bool vis)
+DisplayError_e Display_Init()
 {
-    char out[24];
-    sprintf(out, "%s %s,%d%s", CommandStrings[VIS], CommandStrings[comp], vis ? 1 : 0, TERMINATOR);
-
-    BSP_UART_Write(DISP_OUT, out, strlen(out));
-
-    char buf[8];
-    BSP_UART_Read(DISP_OUT, buf);
-    return (IsNextionFailure(buf[0]) ? ERROR : SUCCESS);
+	BSP_UART_Init(DISP_OUT);
+	return Display_Reset();
 }
 
-/*
- * Initialize the Nextion display
- */
-void Display_Init()
+DisplayError_e Display_Send(DisplayCmd_t cmd)
 {
-    char ret[8];
-    for (int i = 0; i < 8; i++)
-        ret[i] = 0;
-    volatile char *x = ret;
-    BSP_UART_Init(DISP_OUT);
-
-    // The display sends 0x88 when ready, but that might be
-    // before we initialize our UART
-    BSP_UART_Read(DISP_OUT, (char *)x);
-    if (ret[0] == 0x88)
-    {
-        volatile int a = 0;
-        while (1)
-            a++;
-    }
-
-    updateIntValue(PAGE, BCO, NEXTION_DARK_GREY);
-}
-
-/**
- * Set the displayed velocity to vel
- */
-ErrorStatus Display_SetVelocity(float vel)
+	char msgArgs[32];
+	if (cmd.compOrCmd != NULL && cmd.op != NULL && cmd.attr != NULL && cmd.numArgs == 1)
+	{ // Assignment commands have only 1 arg, an operator, and an attribute
+		if (cmd.argTypes[0] == INT_ARG)
+		{
+			sprintf(msgArgs, "%d", (int)cmd.args[0].num);
+		}
+		else
+		{
+			if (cmd.args[0].str == NULL)
+				return DISPLAY_ERR_PARSE;
+			sprintf(msgArgs, "%s", cmd.args[0].str);
+		}
+
+		BSP_UART_Write(DISP_OUT, cmd.compOrCmd, strlen(cmd.compOrCmd));
+		BSP_UART_Write(DISP_OUT, ".", 1);
+		BSP_UART_Write(DISP_OUT, cmd.attr, strlen(cmd.attr));
+		BSP_UART_Write(DISP_OUT, cmd.op, strlen(cmd.op));
+	}
+	else if (cmd.op == NULL && cmd.attr == NULL)
+	{					  // Operational commands have no attribute and no operator, just a command and >= 0 arguments
+		msgArgs[0] = ' '; // No args
+		msgArgs[1] = '\0';
+		if (cmd.numArgs > MAX_ARGS)
+			return DISPLAY_ERR_OTHER;
+		if (cmd.numArgs >= 1)
+		{ // If there are arguments
+			for (int i = 0; i < cmd.numArgs; i++)
+			{
+				char arg[16];
+				if (cmd.argTypes[i] == INT_ARG)
+				{
+					sprintf(arg, "%d", (int)cmd.args[i].num);
+				}
+				else
+				{
+					sprintf(arg, "%s", cmd.args[i].str);
+				}
+
+				strcat(msgArgs, arg);
+
+				if (i < cmd.numArgs - 1)
+				{ // delimiter
+					strcat(msgArgs, ",");
+				}
+			}
+		}
+		BSP_UART_Write(DISP_OUT, cmd.compOrCmd, strlen(cmd.compOrCmd));
+	}
+	else
+	{ // Error parsing command struct
+		return DISPLAY_ERR_PARSE;
+	}
+
+	if (cmd.numArgs >= 1)
+	{ // If there are arguments
+		BSP_UART_Write(DISP_OUT, msgArgs, strlen(msgArgs));
+	}
+
+	BSP_UART_Write(DISP_OUT, (char *)TERMINATOR, strlen(TERMINATOR));
+
+	char buf[8];
```

Comment:
why are we writing 8 garbage bytes? If the idea is to transmit 8 bytes of 0s, maybe initialize it specifically ie. `char buf[8] = {0};`

Reply:
This was something we did during testing, but it's not required so I'll fix it.